### ISSUE TITLE
Wont compile on CentOS 7

### ISSUE THREAD

**Issue by dovi5988** [lwfinger/rtlwifi_new#85]
Hi,

It wont allow me to build on CentOS 7

[root@localhost rtlwifi_new-master]# make
make -C /lib/modules/3.10.0-327.3.1.el7.x86_64.debug/build M=/usr/src/rtlwifi_new-master modules
make[1]: Entering directory `/usr/src/kernels/3.10.0-327.3.1.el7.x86_64.debug'
  CC [M]  /usr/src/rtlwifi_new-master/pci.o
In file included from /usr/src/rtlwifi_new-master/pci.c:33:0:
/usr/src/rtlwifi_new-master/pci.h:349:1: error: redefinition of â€˜pci_zalloc_consistentâ€™
 pci_zalloc_consistent(struct pci_dev *hwdev, size_t size,
 ^
In file included from /usr/src/kernels/3.10.0-327.3.1.el7.x86_64.debug/arch/x86/include/asm/pci.h:118:0,
                 from include/linux/pci.h:1449,
                 from /usr/src/rtlwifi_new-master/pci.h:33,
                 from /usr/src/rtlwifi_new-master/pci.c:33:
include/asm-generic/pci-dma-compat.h:23:1: note: previous definition of â€˜pci_zalloc_consistentâ€™ was here
 pci_zalloc_consistent(struct pci_dev *hwdev, size_t size,
 ^
make[2]: *** [/usr/src/rtlwifi_new-master/pci.o] Error 1
make[1]: *** [_module_/usr/src/rtlwifi_new-master] Error 2
make[1]: Leaving directory`/usr/src/kernels/3.10.0-327.3.1.el7.x86_64.debug'
make: **\* [all] Error 2
[root@localhost rtlwifi_new-master]#

**Comment by lwfinger** [comment]
The code is correct. That function was first entered into the kernel with version 3.17. The source in pci.h says

#if (LINUX_VERSION_CODE < KERNEL_VERSION(3,17,0))
static inline void *
pci_zalloc_consistent(struct pci_dev *hwdev, size_t size,
                      dma_addr_t *dma_handle)
{
        void *tmp;

        tmp = pci_alloc_consistent(hwdev, size, dma_handle);
        if (tmp)
                memset(tmp, 0, size);
        return tmp;
}
#endif

My only conclusion is that CentOS has backported that routine for some other reason. As it is not possible for me to handle every such change created by every distro, you will need to modify your source. You could delete all the code I posted above, or change the "17" in the first line to "10".

**Comment by saraAlizadeh** [comment]
hey.I have the same issue. I changed the "17" to "10" and ran `make` the I got the following message:

`make -C /lib/modules/3.10.0-327.4.5.el7.x86_64/build M=/home/admin/rtlwifi_new modules
make[1]: Entering directory `/usr/src/kernels/3.10.0-327.4.5.el7.x86_64'
  CC [M]  /home/admin/rtlwifi_new/pci.o
  CC [M]  /home/admin/rtlwifi_new/base.o
/home/admin/rtlwifi_new/base.c: In function ‘_rtl_init_mac80211’:
/home/admin/rtlwifi_new/base.c:489:21: error: request for member ‘flags’ in something not a structure or union
    hw->wiphy->wowlan.flags = WIPHY_WOWLAN_MAGIC_PKT;
                     ^
/home/admin/rtlwifi_new/base.c:491:21: error: request for member ‘n_patterns’ in something not a structure or union
    hw->wiphy->wowlan.n_patterns =
                     ^
/home/admin/rtlwifi_new/base.c:493:21: error: request for member ‘pattern_min_len’ in something not a structure or union
    hw->wiphy->wowlan.pattern_min_len =
                     ^
/home/admin/rtlwifi_new/base.c:495:21: error: request for member ‘pattern_max_len’ in something not a structure or union
    hw->wiphy->wowlan.pattern_max_len =
                     ^
make[2]: *** [/home/admin/rtlwifi_new/base.o] Error 1
make[1]: *** [_module_/home/admin/rtlwifi_new] Error 2
make[1]: Leaving directory `/usr/src/kernels/3.10.0-327.4.5.el7.x86_64'
make: *** [all] Error 2
`

after that I tried deleting the part of code you have specified and then I got this message: 
`make -C /lib/modules/3.10.0-327.4.5.el7.x86_64/build M=/home/admin/rtlwifi_new modules
make[1]: Entering directory `/usr/src/kernels/3.10.0-327.4.5.el7.x86_64'
  CC [M]  /home/admin/rtlwifi_new/pci.o
In file included from /home/admin/rtlwifi_new/pci.c:33:0:
/home/admin/rtlwifi_new/pci.h:349:1: error: redefinition of ‘pci_zalloc_consistent’
 pci_zalloc_consistent(struct pci_dev *hwdev, size_t size,
 ^
In file included from /usr/src/kernels/3.10.0-327.4.5.el7.x86_64/arch/x86/include/asm/pci.h:118:0,
                 from include/linux/pci.h:1449,
                 from /home/admin/rtlwifi_new/pci.h:33,
                 from /home/admin/rtlwifi_new/pci.c:33:
include/asm-generic/pci-dma-compat.h:23:1: note: previous definition of ‘pci_zalloc_consistent’ was here
 pci_zalloc_consistent(struct pci_dev *hwdev, size_t size,
 ^
In file included from /home/admin/rtlwifi_new/pci.c:33:0:
/home/admin/rtlwifi_new/pci.h: In function ‘pci_zalloc_consistent’:
/home/admin/rtlwifi_new/pci.h:352:8: warning: unused variable ‘tmp’ [-Wunused-variable]
  void *tmp;
        ^
/home/admin/rtlwifi_new/pci.h:355:1: warning: no return statement in function returning non-void [-Wreturn-type]
 }
 ^
make[2]: *** [/home/admin/rtlwifi_new/pci.o] Error 1
make[1]: *** [_module_/home/admin/rtlwifi_new] Error 2
make[1]: Leaving directory `/usr/src/kernels/3.10.0-327.4.5.el7.x86_64'
make: *** [all] Error 2
`

any idea how to solve it? I couldn't fine anything by googling.

**Comment by lwfinger** [comment]
I would not expect that you would find this error on Google. As I said before, this is a problem introduced by your distro. As it appears that neither of you with this problem is particularly adept in kernel coding, could either of you make your machine available for me to login via SSH? For me to setup a Centos 7 system would take more time that I have, but I could find one hour for this problem.

**Comment by saraAlizadeh** [comment]
unfortunately SSH is not possible to me because I don't have a public IP. is there any other way?I can post you command results. Know this will take time but I desperately need this.

**Comment by kenjeffries** [comment]
For CentOS 7.1 kernel 3.10.0-229.11.1.el7.x86_64 I am able to do a clean make after making these changes:

[change.txt](https://github.com/lwfinger/rtlwifi_new/files/213695/change.txt)

This is on a small Asus VC62B-B013M with built in RTL8821AE. I was able to get the Realtek to authenticate to an access point only after rebooting. I suspect that the correct "systemctl restart"s would have been sufficient. I am still confused about getting WiFi going. There seems to be three separate but overlapping mechanisms in play: wpa_supplicant, NetworkManager including NetworkManager-wifi and the straight forward ifconfig files in /etc/sysconfig/network-scripts/. YMMV.

**Comment by lwfinger** [comment]
Those changes are needed because CentOS is backporting API changes. According to the rules of Linux, that is not allowed. In any case, I will NOT be modifying the source to handle such changes as separate code would be needed for nearly every distro.

As to the other point, wpa_supplicant is used in encryption and decryption of WPA and WPA2 security mode, but that is automatically handled by NetworkManager. If you choose not to run NM, then the network scripts would come into play, otherwise you should be able to ignore them. If you still have problems, then the CentOS forums would be the best place to ask questions. The only question would be how old a version of NM that is in use.

**Comment by kenjeffries** [comment]
Hi lwfinger, yes. Your earlier comments about distros and back porting were key. As I went through the make errors and by looking at the linux cross references for the various kernel versions it was obvious that what you said about the back porting was exactly right. I offered the patch set, not at all as a suggestion for you to incorporate but just as an aide for someone like me that wanted to get the code working on CentOS 7.1. 

Prior to CentOS 7, I always disabled NM. (None of my systems were mobile and NM just caused trouble.) Last year when I started trying to get WiFi to work with 7.1 (using a couple of different USB/WiFi dodads) I discovered that NM had been split up and AFAICT, NM-WiFi was required to get WiFi to work. Now that may not be true but it was my conclusion at the time.

**Comment by mtahamasood** [comment]
@kenjeffries  can you please mention the command with which to apply the patch mentioned in the change.txt you have attached.  doing a "patch change.txt" on the folder which contains rtlwifi_new does not work ( patch sleeps indefinitely)
thanks

**Comment by lwfinger** [comment]
A completely new version of the drivers has just been pushed. It fixes a lot of things. As a result, I am closing all open issues. If you get a new one, please create a new issue.

Note: The new driver fixes the problems with power save. It is recommended that you clear all ips, swlps, and fwlps options.
